### PR TITLE
feat(text_tasks): add external LM captioning helpers

### PR DESCRIPTION
**Pull request by 1larity**
## Summary
This PR adds the small helper module used to build external-LM formatting intents, detect low-value echoed captions, and apply local fallback metadata shaping.

## Scope
In scope:
- caption retry heuristics
- local metadata override helpers
- fallback caption generation
- format-intent string construction
- focused unit tests for the helper module

Out of scope:
- runtime credential access
- HTTP request execution
- task orchestration
- UI wiring
- docs

## What Changed
- add `external_lm_captioning.py` for formatting-intent construction and local caption fallback helpers
- add `external_lm_captioning_test.py` to cover retry heuristics, metadata overrides, fallback captions, and format-intent assembly

## Testing
```bash
/media/stellaris/Linux AI/Acestep/acestep_git/ACE-Step-1.5/ACE-Step-1.5/.venv/bin/python -m unittest \
  acestep.text_tasks.external_lm_captioning_test
```

Ran 4 tests, all passing.

## Risk / Compatibility
This PR is based directly on `main`. It stays inside a single helper module under `acestep/text_tasks` and does not introduce live provider requests or secret handling.

## Related Upstream Context

This PR is one small slice of the broader external-LM work that was originally bundled into [#808](https://github.com/ace-step/ACE-Step-1.5/pull/808).

Related upstream references:
- [#717](https://github.com/ace-step/ACE-Step-1.5/issues/717) Running LM models via ollama
- [#808](https://github.com/ace-step/ACE-Step-1.5/pull/808) Legacy GPU stability, external AI generation, and better progress UX
- [#677](https://github.com/ace-step/ACE-Step-1.5/issues/677) "Enhance Caption" is worse than useless!
- [#681](https://github.com/ace-step/ACE-Step-1.5/issues/681) "AI Auto labelling" hallucinates lyrics
- [#472](https://github.com/ace-step/ACE-Step-1.5/issues/472) lora auto-label metadata is horrible

This slice focuses only on captioning helper behavior and metadata handling. It does not add external provider transport, request execution, or UI wiring.

## Summary by CodeRabbit

* **New Features**
  * Caption-processing utilities: caption normalization, retry heuristics, user-metadata overrides (with value coercion and alias synchronization), fallback caption generation, and assembling formatted requests that include caption, lyrics, and relevant metadata.
* **Tests**
  * Added tests validating retry behavior, metadata override/alias propagation, fallback caption content, and formatted-request output (including omission of empty/unknown metadata).

### CODE REVIEW

**Comment by 1larity** [comment]
Reviewed the @CodeRabbit findings against the current branch.

Applied in `e0297dc`:
- fixed `apply_user_metadata_overrides()` so it writes both field-name variants for key scale, time signature, and language
- updated the captioning test fixture to use the canonical runtime keys instead of the alias-only names
- added `loguru.logger.debug()` logging for ignored invalid `bpm` / `duration` metadata conversions

Verified as not applicable to the current code:
- the `init_service_test.py` merge-marker report does not match the current tree; the file is 758 lines long, contains no `<<<<<<<` markers, and has a single `test_load_text_encoder_uses_cpu_safe_dtype_when_offloaded` definition at line 715

Validation rerun:
```bash
/media/stellaris/Linux AI/Acestep/acestep_git/ACE-Step-1.5/ACE-Step-1.5/.venv/bin/python -m unittest \
  acestep.text_tasks.external_lm_captioning_test
```

Ran 4 tests, all passing.

File: acestep/text_tasks/external_lm_captioning_test.py [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    def test_apply_user_metadata_overrides_preserves_constrained_values(self) -> None:
+        """User-supplied metadata should win over provider drift."""
+
+        plan = SimpleNamespace(
+            bpm=1,
+            duration=2.4,
+            key_scale="C minor",
+            time_signature="3/4",
+            vocal_language="English",
+        )
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/text_tasks/external_lm_captioning_test.py:54]: Fixed in e0297dc. I verified this one against the current branch: the original test fixture was masking the real runtime field names. The fixture now uses canonical `keyscale` / `timesignature` / `language`, and the assertions check the canonical fields directly while still covering the compatibility aliases updated in the helper.

File: acestep/text_tasks/external_lm_captioning.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+        try:
+            plan.duration = float(user_metadata["duration"])
+        except (TypeError, ValueError):
+            pass
+    if user_metadata.get("keyscale"):
+        plan.key_scale = str(user_metadata["keyscale"]).strip()
+    if user_metadata.get("timesignature"):
+        plan.time_signature = str(user_metadata["timesignature"]).strip()
+    if user_metadata.get("language"):
+        plan.vocal_language = str(user_metadata["language"]).strip()
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/text_tasks/external_lm_captioning.py]: Fixed in e0297dc. I verified this one against the current code and agreed with the finding. `apply_user_metadata_overrides()` now applies both common field variants for key scale, time signature, and language so user constraints stick regardless of whether the plan object uses `keyscale`/`timesignature`/`language` or `key_scale`/`time_signature`/`vocal_language`.

File: acestep/text_tasks/external_lm_captioning.py [hunk 3]
```diff
@@ -0,0 +1,106 @@
...
+        language = str(user_metadata["language"]).strip()
+        plan.language = language
+        plan.vocal_language = language
+    return plan
+
+
+def build_fallback_caption(*, caption: str, user_metadata: dict[str, Any]) -> str:
+    """Build a simple local narrative fallback when the provider keeps echoing input."""
+
+    source = (caption or "music piece").strip().rstrip(".")
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/text_tasks/external_lm_captioning.py]: Fixed in `0e00321`: `build_fallback_caption()` now falls back to `"music piece"` after stripping, so whitespace-only captions no longer produce a broken opening sentence. I also added a regression test covering the whitespace-only case.

**Comment by 1larity** [comment]
Addressed the two new `#882` review findings in `0e00321`.

Included updates:
- whitespace-only captions now fall back to `"music piece"` in `build_fallback_caption()`
- string metadata values are normalized before the `unknown` filter in `build_format_request_intent()`
- added regression coverage for both behaviors in `external_lm_captioning_test.py`

- Validation:
- `python -m unittest acestep.text_tasks.external_lm_captioning_test`
- 6 tests passed

File: acestep/text_tasks/external_lm_captioning.py [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+
+    intent_parts = [
+        "Please format and enrich the following for ACE-Step generation.",
+        f"Caption: {caption or ''}",
+        f"Lyrics: {lyrics or ''}",
+    ]
+    for key in ("bpm", "duration", "keyscale", "timesignature", "language"):
+        value = user_metadata.get(key)
+        if value not in (None, "", "unknown"):
+            intent_parts.append(f"{key}: {value}")
```
[the remark this answers is not in the document]
Comment (1larity) [reply acestep/text_tasks/external_lm_captioning.py:101]: Fixed in `0e00321`: string metadata values are now stripped before the `unknown` check, so variants like `"Unknown"` and `" unknown "` are filtered out while non-string values still pass through unchanged. There is also regression coverage for the normalized string case.

**Comment by 1larity** [comment]
Reviewed the latest `#882` notes against the current branch and took the remaining captioning-level fixes that still felt worth landing in-scope for this slice.

Fixed in `9959199e` and `8fc7d193`:
- `build_format_request_intent()` now filters a slightly broader set of placeholder metadata sentinels (`unknown`, `none`, `n/a`, `default`, plus empty strings) before building the provider-facing intent.
- `build_fallback_caption()` is no longer English-only for the currently supported app languages. The localized fallback prose now lives in a small helper module and switches on the requested language code for `en`, `ja`, `zh`, and `he`.
- added focused regression coverage for the placeholder filtering and the localized fallback path.

Validation rerun:
```bash
/media/stellaris/Linux\ AI/Acestep/acestep_git/ACE-Step-1.5/ACE-Step-1.5/.venv/bin/python -m unittest \
  acestep.text_tasks.external_lm_captioning_test
```
8 tests passed.

Notes on the remaining review points:
- I did not change the dual attribute writes in `apply_user_metadata_overrides()`. They are an intentional compatibility shim for plan objects that still disagree on field naming, and they are correct for the current state of the runtime.
- I did not relax the `caption_needs_retry()` floor. That threshold came out of CER profiling[^cer] and is behaving usefully in practice, so I’d rather keep it evidence-led and revisit only if the case pack starts showing bad retry behavior.
- I still did not add a dedicated `normalized_caption()` test in this pass; it remains covered indirectly through `caption_needs_retry()`, and I kept this PR focused on behavior changes rather than test-only expansion.

One explicit tradeoff note: I kept the localization small by containing it to the helper layer instead of threading these strings through the full Gradio i18n surface. That still adds a bit of file churn for a tiny PR, and the non-English canned fallback strings may deserve native-speaker review later if they read awkwardly in practice.

[^cer]: CER here is the small case-based caption-enhancement review/profiling workflow I’ve been using around the PR808 quality follow-ups. In practice, that means checking real prompt/result pairs, watching for echo-like or under-expanded rewrites, and only changing the retry heuristic when the case pack shows it is helping rather than just “feeling nicer” in the abstract.

**Review by ChuxiJ — COMMENTED** [review]
## Code Review: feat(text_tasks): add external LM captioning helpers

Captioning helpers for generating captions via external LM providers.

### Strengths
- **Locale fallback**: `external_lm_captioning_fallback_locale.py` handles locale-aware caption generation gracefully
- **Clean API surface**: Simple function-based interface for captioning

### Issues / Suggestions

1. **Prompt injection risk**: If user-provided lyrics or captions are interpolated into prompts sent to external LMs, ensure proper escaping or sandboxing to prevent prompt injection via crafted lyrics.

2. **Timeout handling**: External LM captioning calls should have configurable timeouts — a hung API call could block the generation pipeline.

3. **Fallback behavior**: Document what happens when the external LM returns an empty or nonsensical caption — does it fall back to the original user caption?

Looks good overall.

**Review by ChuxiJ — COMMENTED** [review]
## Follow-up: Critical bug found in deeper analysis

### Bug: `.format()` crash on captions containing `{` or `}`

In `external_lm_captioning_fallback_locale.py:65`:
```python
source = (caption or "").strip().rstrip(".") or template["default_source"]
# ...
template["intro"].format(source=source)
```

If a user's caption contains `{` or `}` characters (e.g., `"Track {remix} edition"`), Python's `.format()` will raise `KeyError` or `IndexError` at runtime.

**Fix**: Replace `.format(source=source)` with `.replace("{source}", source)` for safety, or escape braces in the source string first.

### Additional findings:

- **`caption_needs_retry` false positives**: Short original captions (1-3 words) will always trigger retry. E.g., original=`"Jazz"` (1 word), generated=`"A smooth jazz piano piece"` (5 words) → threshold is `max(6, 1+3) = 6`, so 5 < 6 = retry triggered unnecessarily.

- **`int("120.0")` for BPM**: In `apply_user_metadata_overrides`, `int()` on a decimal string like `"120.0"` raises `ValueError`. Consider `int(float(...))` instead.

- **Missing test coverage**: No tests for error branches, None inputs, brace-containing captions, or zh/he locale fallbacks.

**Comment by 1larity** [comment]
@ChuxiJ 
Thanks for the careful pass. I updated the branch against current `main` and took the helper-level fixes that still felt in-scope for this slice.  
  
On the brace concern: I manually tested a caption containing literal braces in the UI and also added a focused regression case for a caption like `Track {remix} edition`. I wasn’t able to reproduce a `.format()` crash in this path. In this helper, Python is formatting the template string, not reparsing braces inside the substituted caption value.  
  
That manual pass did uncover a separate real bug, though: the helper could crash when `user_metadata` arrived as `None`. I fixed that at the helper boundary and added regression coverage for the missing-metadata case.  
  
 I also took the decimal BPM edge case, so integer-like strings such as `"125.0"` now apply cleanly in the metadata override shim.  
  
On fallback behavior: I clarified the helper docstring so the current behavior is explicit. Format mode retries once when the provider caption is empty, unchanged, or still too short, and then falls back to a local narrative caption built from the original user caption plus metadata. I’d still like to leave the fuller user-facing explanation for the docs pass.  
  
On the timeout point: I agree with the concern in principle, but I didn’t widen this small helper PR to add a new timeout control. Ollama cold-start behavior is model/runtime dependent, and I’d rather expose or tune that in the next slice, where the external-LM flows start becoming usable in a limited but more complete way.  
  
On the prompt-injection note: I’m not treating that as an in-scope security bug in this helper layer. External LM mode is intentionally sending user-provided caption/lyrics to a provider the same user configured with their own credentials, so this reads more like a trust-boundary/documentation concern than a local ACE-Step execution surface.   Why would a user use Ace-step to prompt inject an API they presumably have legitimate access to already, that implies CURL level access already.
  
 I reviewed the `caption_needs_retry` short-prompt heuristic as well, but left it unchanged for this slice. That threshold came from CER-guided tuning, and I’d rather only retune it if the case pack shows it is actually hurting outcomes in practice.  I am planning on running a multi thousand test cases using the CER tool across multiple models and providers (current guidance used 100 unique prompts to about 20 models with 50  prompts each). I really want to do all of the quality analysis on all LM tasks at once so I can control literal inference costs :) I can see needing a month's worth of OpenAi Pro to run all the analysis :P
  
 Test coverage on the branch now includes brace-containing captions, missing metadata inputs, and `zh`/`he` locale fallbacks in addition to the existing retry and placeholder-metadata cases.
